lib/perl/Genome/Model/Tools/Bsmap/methylation_ratio.py

```diff
@@ -0,0 +1,181 @@
+'''
+Created on Jan 26, 2015
+
+@author: ygindin
+'''
+import optparse
+import logging
+import time
+import tempfile
+import pysam
+import os
+import pdb
+from multiprocessing import Process, Queue
+
+def write_data(result_queue, out_file, work_units):
+    start = time.clock()
+    fout = open(out_file, 'w')
+    fout.write('chr\tpos\tratio\ttotal_C\tmethy_C\tunmeth_C\n')
+    tab = '\t'
+    new_line = '\n'
+    print 'writer opens shop. waiting for ', work_units, ' work units'
+    work_written = 0
+    logging.basicConfig(level=logging.DEBUG, format='(%(threadName)-10s) %(message)s')
+
+    while work_written < work_units:
+        counter = result_queue.get()
+        work_written += 1
+        for coordinate in counter.keys():
+            depth = float(counter[coordinate].getCoverage())
+            meth_depth = float(counter[coordinate].getMethylationCoverage())
+            un_methylated = float(counter[coordinate].get_un_methylation_coverage())
+            if meth_depth + un_methylated == 0:
+                # it's a mismatch, not a methylation eligible event
+                continue
+            ratio = meth_depth / (meth_depth + un_methylated)
+            chromosome = counter[coordinate].get_chromosome()
+            
+            fout.write(tab.join([chromosome, str(coordinate+1), str(ratio), str(depth), str(meth_depth), 
+                                 str(un_methylated) + new_line]))            
+        
+        logging.debug('wrote %s', ''.join(['work unit ', str(work_written), ' out of ', str(work_units)]))
+    print 'writer calling quits'
+    fout.close()
+    end = time.clock()
+    run_time = end - start
+    print 'Running time: ', run_time, ' seconds'
+
+def isOnWatsonStrand(tags):
+    """ Parses the ZS field to determine if the read is on the Crick strand """
+    for tag in reversed(tags):
+        if 'ZS' in tag:
+            if tag[1].startswith('+'):
+                return True
+            else:
+                return False
+    return None
+
+
+def process_region_pileup(i, bam_file_path, fasta_path, work_queue, out_queue):
+    bam_file = pysam.AlignmentFile(bam_file_path, 'rb')
+    logging.basicConfig(level=logging.DEBUG,
+                        format='(%(threadName)-10s) %(message)s')
+
+    genome = pysam.FastaFile(fasta_path)
+    nuc_A = 'A'
+    nuc_T = 'T'
+    nuc_C = 'C'
+    nuc_G = 'G'
+    cg = ['C', 'G']
+    while True:
```

Comment:
Where does this terminate?  I don't see a `break`.  Is there another way to escape (I don't know Python well)?